File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
# computers = [
#     [1,1,0],
#     [1,1,0],
#     [0,0,1]
# ]
from collections import deque

def solution(n, computers):
    numberOfNetworks = 0
    
    visited = [False]*n
    
    def bfs(i):
        visited[i] = True
        que = deque()
        que.append(i)
        while que:
            cur_v = que.popleft()
            for j in range(len(computers[cur_v])):
                #if cur_v==j: continue
                if computers[cur_v][j]==1 and visited[j]==False:
                    visited[j] = True
                    que.append(j)
    
    for i in range(n):
        if visited[i]==False:
            bfs(i)    
            numberOfNetworks += 1
    
    return numberOfNetworks

File: 프로그래머스/3/43162. 네트워크/네트워크.java
class Solution {
    public boolean[] visited;
    public int solution(int n, int[][] computers) {
        int answer = 0;

        visited = new boolean[n];
        for (int i = 0; i < n; i++) {
            if (!visited[i]) {
                dfs(i, computers);
                answer++;
            }
        }

        return answer;
    }

    public void dfs(int st, int[][] computers) {
        visited[st] = true;

        for (int i = 0; i < computers[st].length; i++) {
            if (st == i) {
                continue;
            }

            if (computers[st][i] == 1 && !visited[i]) {
                dfs(i, computers);
            }
        }
    }
}